src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -393,6 +397,8 @@ public bool EndsWith(string text, StringComparison comparisonType)
         /// <paramref name="offset"/> or <paramref name="length"/> is less than zero, or <paramref name="offset"/> + <paramref name="length"/> is
         /// greater than <see cref="Length"/>.
         /// </exception>
+        [MemberNotNull(nameof(Buffer))]
+        [MemberNotNull(nameof(Value))]
```

Comment:
The method doesn't make either property non-`null`; it just `throw`s when either are `null` and there's no `[DoesNotReturnIfMemberNull(...)]` attribute available. Suggest leaving the metadata here alone.

Same for the other `Substring(...)` overload.

Reply:
`MemberNotNull` also acts as `DoesNotReturnIfMemberNull` as I described in https://github.com/dotnet/runtime/pull/58308#discussion_r697905372

It's similar to `[NotNull]`. If method throws argument is still considered as nullable. Example: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L59